ENH: signal: Compute envelope of a real- or complex-valued signal

DietBru opened the pull request:
This PR proposes a new function for `scipy.signal`, which computes the envelope of a real- or complex-valued signal after applying an optional bandpass filter.

The discussion in #19814 revealed that literature is a bit convoluted regarding on how to define envelopes for signals. Hence, this PR also tries to concisely document the approach utilized here.

This PR differs from #19814 in that it provides a general solution, whereas #19814  does not produce correct results (regarding the definitions of a envelope given here) for all possible signals.

DietBru committed:
ENH: Compute envelope of a real- or complex-valued signal.

This PR proposes a new function for `scipy.signal`, which computes the envelope of a real- or complex-valued signal after applying an optional bandpass filter.

Closes #19814.

DietBru commented:
The rendered version can be found [here](https://output.circle-artifacts.com/output/job/4e138932-2102-4506-bc3a-b634a2085c18/artifacts/0/html/reference/generated/scipy.signal.envelope.html#scipy.signal.envelope).

I am not sure what to make of the error message:
```
scipy/signal/_signaltools.py:2467: in <module>
    def envelope(z: np.ndarray, bp_in: tuple[int | None, int | None] = (1, None), *,
E   TypeError: unsupported operand type(s) for |: 'type' and 'NoneType'
```
This is the only error I found in the failing tests which could be related this PR.

lucascolley commented:
That syntax is not supported in Python < 3.10. `from __future__ import annotations` at the top of the file might work. Alternatively, keeping the typing in a stub `.pyi` file may be easier to work with for now. Alternatively, you can use `Union`, however, I think the linter may complain about that (not sure).

DietBru commented:
Thanks @lucascolley, for the prompt answer. It could have been that I asked this question before :sweat_smile:

DietBru committed:
Add ``from __future__ import annotations`` to supress warnings in CI.

j-bowhay commented:
@yagizolmez would you be interested in reviewing this?

yagizolmez commented:
@j-bowhay I will leave a review in the upcoming days. I want to note that this is NOT a competing PR with #19814. Following the discussion we had there, I decided to break up the envelope function I proposed and not implement the analytical envelope.

@DietBru I  have a quick question before I leave a more detailed review. Comparing this function with `hilbert`, major additions seem to be:

1. Support for complex-valued signals.
2. The option to upsample the envelope.
3. The option to apply a bandpass filter.

Have you considered extending `hilbert` rather than adding a new function?

DietBru commented:
> Have you considered extending hilbert rather than adding a new function?

I did, but I came to the conclusion that that would only lead to confusion:

The [analytic signal](https://en.wikipedia.org/wiki/Analytic_signal#Complex_envelope/No) is a concept based on  real-valued signals. Rightfully, [hilbert()](https://docs.scipy.org/doc/scipy/reference/generated/scipy.signal.hilbert.html) raises a `ValueError` for complex-valued inputs. Also, changing that behavior could break existing implementations.

 The envelope on the other hand is a concept that is inherent to complex-valued signals with real-valued signals being a special case. Note that the treatment of  real-valued signals depends on the underlying assumptions of what the signal represents. This is illustrated in the second example of the `envelope()`'s doc string.

yagizolmez commented:
@DietBru I am sorry for the delay. I have been very busy these days, so I am not yet able to provide a detailed review. One thing catched my eye. Under certain circumstances, this function must give the same result with `hilbert()`, i.e. when `z` is real and `bp_in` is selected appropriately. Have you considered writing some tests and benchmarks covering these situations?

DietBru committed:
Refactoring to
 * obtain identical results as with `hilbert()`,
 * some polishing in the docstring,
 * add "See Also" to `hilbert()`.

DietBru commented:
> One thing catched my eye. Under certain circumstances, this function must give the same result with `hilbert()`, ...

Good catch, thank you. I found an edge case where the results differ. Hence, I did some refactoring to ensure that `envelope()` and `hilbert()` always behave identically. I think it should ready for complete review now.

lucascolley commented:
@yagizolmez would you like to take another look at this?

yagizolmez commented:
@lucascolley Thanks for reminding me Lucas. I have gotten extremely busy in the school lately. I will take a look when I have time. @DietBru I apologize.

DietBru commented:
Resolved merge conflict. Current version of the rendered documentation can be found [here].

[here]: https://output.circle-artifacts.com/output/job/512a8bbd-58c9-4c88-a894-2b4ac9a48dbd/artifacts/0/html/reference/generated/scipy.signal.envelope.html

tylerjereddy commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -2800,6 +2800,187 @@ def test_hilbert2_types(self, dtype):
+    @classmethod
+    def raiseValErr(cls, match):
+        """Little helper to raise ValueError for mor concise code layout. """
+        return pytest.raises(ValueError, match=match)
```
`mor` ->` `more` in the docstring above.

I wonder if we really want this shim though, the kwarg incantation of the function was obviously intended on the `pytest` side, and that's how we're using it elsewhere in SciPy. That said, pretty minor point!

DietBru replied:
Thanks for taking a look, @tylerjereddy. Regarding the shim, I guess it is matter of personal taste. I don't really have strong a opinion on this. I removed the shim in commit [cbdc3ed](https://github.com/scipy/scipy/pull/20097/commits/cbdc3edd2c848118826dd6e3d082419da281877d).

tylerjereddy reviewed (commented):
I read through the discussion here.

This looks like a decent chunk of work, but I'll have to bump the release milestone since there's no detailed expert review of the new feature. I could check it for superficial things like test coverage and so on, but I wouldn't feel comfortable merging it on my own. Apologies!

DietBru committed:
Removed `pytest.raises` helper in test_signaltools.py

yagizolmez commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2458,271 @@ def hilbert2(x, N=None):
+        The corner frequencies are specified as multiples of `1/(n*T)` and the allowed
+        frequency range is specified by ``-n//2 <= bp_in[0] < bp_in[1] <= (n+1)//2``.
+        ``None`` entries are replaced with `-n//2` or `(n+1)//2` respectively. The
+        default of ``(1, None)`` removes the mean value as well as the negative
+        frequency components.
```
This part confuses me a little bit. For a real-valued signal, it makes sense to remove the negative frequency components. For a complex-valued signal, I don't fully understand.

DietBru replied:
In communications engineering it is common to modulate/demodulate a signal, i.e. shift it along the frequency axis. E.g., one might have shifted a signal into the (complex-valued) base-band for efficiency reasons do some signal processing and shift it afterwards into another frequency band. Hence, it is convenient for some application scenarios to allow arbitrary frequency bands for calculating the envelope.

yagizolmez replied:
Could you share some resources with me? I would like to understand this part a little better.

yagizolmez replied:
Wouldn't the default value here also remove sub 1 Hz frequencies? Maybe you could add an optional parameter for removing the mean value, and do this operation strictly in the time domain?

DietBru replied:
I am not aware of any other literature dedicated to discussing envelopes. Typically it is mingled with the Hilbert transform. Perhaps reading in [Proakis' book] how envelopes are used could give you some insights.

[Proakis' book]: https://books.google.de/books/about/Digital_Communications.html?id=HroiQAAACAAJ

yagizolmez commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2458,271 @@ def hilbert2(x, N=None):
+        ``None`` entries are replaced with `-n//2` or `(n+1)//2` respectively. The
+        default of ``(1, None)`` removes the mean value as well as the negative
+        frequency components.
+    n_out :
+        If not ``None`` the output will be resampled to `n_out` samples.
```
What is the default for n_out? I would suggest making it the same length as the input signal.

DietBru replied:
Added default case in  commit f910d6e7

yagizolmez replied:
I had some second thoughts on this. Would it be better to set the default as no upsampling? This would be consistent with the output of hilbert(). I am undecided though. What do you think? Maybe we could ask for more thoughts on this after I am done reviewing the other parts of the code?

yagizolmez commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2458,271 @@ def hilbert2(x, N=None):
+        returns the contents of the frequency band ``< bp_in[0]``. If ``None`` then
+        nothing is returned.
+    axis :
+       Axis of `z` over which to compute the envelope (default is last axis).
```
```suggestion
       Axis of `z` over which to compute the envelope. Default is the last axis.
```

yagizolmez commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2458,271 @@ def hilbert2(x, N=None):
+
+    * The signal is assumed to be periodic. Discontinuities between signal start and
+      end can lead to unwanted results due to Gibbs phenomenon.
+    * The FFT is slow if the signal length is prime or very long. Also, the memory
+      demands are typically higher than a comparable FIR/IIR filter based
+      implementation.
```
Is there a way to make this implementation robust to large prime numbers? Large prime signal length occurs relatively often.

DietBru replied:
AFAIK, not really. Depending on circumstances, either resampling or zero-padding  (consult [next_fast_len()]) might produce better results. Furthermore, [Lyons] (listed in the reference), provides further ways to calculate the envelope for different kinds of signals. Which approach is best heavily depends on the application.



[next_fast_len()]: https://docs.scipy.org/doc/scipy/reference/generated/scipy.fft.next_fast_len.html
[Lyons]: <https://community.infineon.com/gfawx74859/attachments/gfawx74859/psoc135/46469/1/R.%20Lyons_envelope_detection_v3.pdf>

yagizolmez reviewed (changes requested):
@DietBru Sorry for the delay. In order to keep things moving, I reviewed the documentation part of the code (which may eventually cause changes to the actual code). I haven't reviewed the examples yet, because the docs don't render. I will review the other parts soon.

DietBru commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2458,271 @@ def hilbert2(x, N=None):
+    r"""Compute the envelope of a real- or complex-valued signal after applying an
+    optional bandpass filter.
+
+    Any complex-valued signal `z(t)` can be split into a real-valued (instantaneous)
+    amplitude `a(t)` and a real-valued (instantaneous) phase `phi(t)`, i.e.,
+    `z(t) = a(t) * exp(1j*phi(t))`. The envelope is defined as the absolute value of
```
The redundancy is intentional here.  In signal processing, when talking about phase and amplitude it is often implicitly assumed that those are  stationary. I think, the redundancy makes it easier to grasp for many people.

DietBru replied:
Changed the wording slightly in commit f910d6e7

yagizolmez replied:
I think it would be better to write phi in Greek letters. Could you write the math expressions in latex? If this is not possible, I would suggest denoting the phase with p instead of phi.

DietBru replied:
I also like the LaTeX expression better, but not all IDEs render those correctly when displaying help texts. 

There is guidance on that topic in SciPy, though not very straightforward to find:
The [SciPy contributor Guide] links to the  [Numpy Documentation Style] , which links to [Numpydoc] guide
which states that LaTeX may be used in the [Notes section].

So I followed that convention.

[SciPy contributor Guide]: https://docs.scipy.org/doc/scipy/dev/contributor/contributor_toc.html#documentation
[Numpy Documentation Style]: https://numpy.org/devdocs/dev/howto-docs.html#howto-document
[Numpydoc]: https://numpydoc.readthedocs.io/en/latest/format.html
 [Notes section]: https://numpydoc.readthedocs.io/en/latest/format.html#notes

DietBru replied:
I guess using `p(t)` instead of `phi(t)` is a matter of taste—I'm just used to signal processing literature where an phase angle is always denoted by a Greek letter.

DietBru committed:
Docstrings improvements based on @yagizolmez recommendations.

DietBru commented on scipy/signal/_signaltools.py:
```diff
@@ -2327,6 +2329,10 @@ def hilbert(x, N=None, axis=-1):
 
+    See Also
+    --------
+    envelope: Compute envelope of a signal after applying an optional bandpass filter.
```
Fixed in commit f910d6e7

DietBru committed:
Fixed unrelated Linter error which appeared in CI.

DietBru commented:
Thanks for the review @yagizolmez. The updated rendered version of the documentation can be found [here],


[here]: https://output.circle-artifacts.com/output/job/d225ca93-bccc-4b0b-85d6-7e415f77c080/artifacts/0/html/reference/generated/scipy.signal.envelope.html

yagizolmez commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,269 @@ def hilbert2(x, N=None):
+    projection into the 2d real and imaginary coordinate planes of trajectory and tube
+    is depicted.
+
+    The left plot shows an analytical signal, i.e, the phase difference between
+    real and imaginary is always 90 degrees resulting in a spiraling trajectory. It can
+    be seen that in this case the real part has also the expected envelope, i.e.,
+    representing the absolute value of the instantaneous amplitude. The right plot
+    shows a real-valued signal interpreted as a complex signal, i.e., having zero
+    imaginary part. There the resulting envelope is not as smooth as in the analytic
+    case and the instantaneous amplitude in the real plane is not recovered.
```
I would divide this second example into two. 

The first example can be about the geometric envelope interpretation of a complex-valued signal. You would keep only the plot on the left for this one.

The second example must emphasize that `z(t) = a(t)` and `z(t) = a(t) + 0j` are interpreted differently and will yield different results. You could also possibly combine this example with the example about the real-valued signal you provide earlier.

DietBru replied:
I expanded the description in commit bf9978fe201 to explicitly discuss the difference between a real-valued signal and the same signal with zero imaginary part, which also illustrates the relationship between those two plots. Hence, I think that separating those two plots would not bring any additional clarity npw.

yagizolmez reviewed (changes requested):
@DietBru I did a second pass on the documentation, now that the examples are rendered. Also, I would like to point out one thing. I don't have much experience ever needing to extract the envelope of a complex-valued signal. I can review for math accuracy, but I cannot review for user-friendliness/usefulness etc.

@ilayn Do you know anyone who could provide a better review for the complex-valued signal case?

lucascolley commented:
> Do you know anyone who could provide a better review for the complex-valued signal case?

Perhaps @roryyorke or @larsoner ?

yagizolmez commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,269 @@ def hilbert2(x, N=None):
+            Z[..., bp] *= 2
+        elif bp.stop > 0:
+            Z[..., 1:bp.stop] *= 2
+    if not (bp.start <= 0 < bp.stop):  # envelope is invariant to freq. shifts.
+        z_bb = sp_fft.ifft(Z[..., bp], n=n_out) * fak  # baseband signal
```
Why do we need to shift it otherwise?

DietBru replied:
To save a little bit of computational effort by shortening the `ifft` input

yagizolmez reviewed (commented):
The implementation largely looks good to me. Although I didn't understand some parts. I will start going over the unit tests.

yagizolmez commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,269 @@ def hilbert2(x, N=None):
+        If not ``None`` the output will be resampled to `n_out` samples. The default
+        of ``None`` sets the output to the same length as the input `z`.
+    squared :
+        If set, the square of the envelope is returned. This is useful, if filters need
+        to be applied to the result, since the squared envelope possesses fewer
+        harmonics.
```
Could you please add an example on this? Especially about its usefulness.

DietBru replied:
Let me put some more thought into this.

DietBru replied:
I was not able to come up with meaningful example which could be described in a single paragraph. I changed the description, which mentions now the effect of the square root in the abs function.

yagizolmez commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -2800,6 +2800,188 @@ def test_hilbert2_types(self, dtype):
+        """Test envelope calculation with real-valued test signals.
+
+        The comparisons are performed in the Fourier space, since it makes evaluating
+        the bandpass filter behavior straightforward. Note that also the squared
+        envelope can be easily calculated by hand, if one recalls that coefficients of
```
Not-squared envelope can also be calculated by hand. I think there should be tests covering squared=False too.

DietBru replied:
From a testing perspective, the only difference between a squared and a non-squared envelope is [this line]:
```python
    z_env = np.abs(z_bb) if not squared else z_bb.real ** 2 + z_bb.imag ** 2
```
It is verified once in [test_envelope_verify_parameters()], which should be sufficient, IMHO.

BTW, I am not aware how to easily determine the spectrum of a square-root of a sine wave by hand. Square-rooting introduces an infinite amount of harmonics to each sinusoid.

[this line]: https://github.com/scipy/scipy/blob/2838e7d52b1b623c6c55448fdeb0f98b955d8a1a/scipy/signal/_signaltools.py#L2702
[test_envelope_verify_parameters()]: https://github.com/scipy/scipy/blob/2838e7d52b1b623c6c55448fdeb0f98b955d8a1a/scipy/signal/tests/test_signaltools.py#L2849

yagizolmez commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -2800,6 +2800,188 @@ def test_hilbert2_types(self, dtype):
+        bandpass filter were already tested in the previous test.
+        """
+        z = sp_fft.ifft(sp_fft.ifftshift(Z))
+        ze2, zr = envelope(z, bp_in, residual='all', squared=True)
```
Can you also test for squared=False?

DietBru replied:
Consult [comment above](https://github.com/scipy/scipy/pull/20097/files#r1694233553).

yagizolmez commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -2800,6 +2800,188 @@ def test_hilbert2_types(self, dtype):
+                               dtype=complex)
+        Zr_desired = np.array([[1, 0, 0, 0, 0], [7, 0, 0, 0, 0]], dtype=complex)
+
+        ze2, zr = envelope(z, squared=True, axis=1)
+        ye2T, yrT = envelope(z.T, squared=True, axis=0)
```
Can you also test for squared = False?

yagizolmez commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -2800,6 +2800,188 @@ def test_hilbert2_types(self, dtype):
+                           dtype=complex)
+        Zr_des = np.array([[1, 0, 0, 0, 2], [1, 0, 0, 0, 2]], dtype=complex)
+
+        kw = dict(bp_in=(-1, 2), residual='all', squared=True)
```
Can you also test for squared = False?

DietBru replied:
Consult [comment above](https://github.com/scipy/scipy/pull/20097/files#r1694233553).

yagizolmez reviewed (changes requested):
First pass at the tests looks good. I will take a second look in the near future. I think you could add some tests for squared = False too? This could also be calculated by hand for simple signals.

yagizolmez commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,269 @@ def hilbert2(x, N=None):
+    The following plot illustrates the envelope of a signal with variable frequency and
+    a low-frequency drift. To separate the drift from the envelope, a 4 Hz highpass
+    filter is used. The low-pass residuum of the input bandpass filter is utilized to
+    determine an asymmetric upper and lower bound to enclose the signal. Due to the
+    smoothness of the resulting envelope, it is down-sampled from 1000 to 40 samples.
```
My bad. The current default is consistent with hilbert.

DietBru committed:
Update scipy/signal/_signaltools.py

Remove unrelated whitespaces

Co-authored-by: Yagiz Olmez <yagiz.olmez@gmail.com>

DietBru committed:
Update scipy/signal/_signaltools.py

Co-authored-by: Yagiz Olmez <yagiz.olmez@gmail.com>

DietBru committed:
Improved docstring: Incorporated @yagizolmez's suggestions and smoothed out some edges.

DietBru commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,269 @@ def hilbert2(x, N=None):
+    >>> from scipy.signal.windows import gaussian
+    >>> from scipy.signal import envelope
+    ...
+    >>> n, T, n_out = 1000, 1/1000, 40
```
Good point! Fixed in commit bf9978fe.

DietBru commented:
The [failure] in building the docs seem to be a `jupyterlite_sphinx` problem, hence, unreleated to this PR—will look into that later on.

[failure]: https://app.circleci.com/pipelines/github/scipy/scipy/29226/workflows/3adf6a71-c5eb-4a0f-86da-07ab51ecd2b3/jobs/98620?invite=true#step-104-10027_34

lucascolley commented:
Doc build failure is tracked in gh-21323

DietBru commented:
Rebased this branch on the documentation is building again. The current `envelope` documentation can be found [here]. The [test failures] are generated by `special/tests/test_orthogonal_eval.py`, hence unrelated to this PR.

Thanks again for your input @yagizolmez — I hope I have sufficiently addressed all your remarks by now.



[test failures]: https://github.com/scipy/scipy/actions/runs/10339348990/job/28618581398#step:8:1692
[here]: https://output.circle-artifacts.com/output/job/220d15f8-df3f-4b5e-b51a-f980c53d3003/artifacts/0/html/reference/generated/scipy.signal.envelope.html

yagizolmez commented:
@DietBru Thanks. I will check in a few days.

yagizolmez reviewed (approved):
Looks good to me! I would suggest getting at least one more opinion before merging though.

j-bowhay commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,277 @@ def hilbert2(x, N=None):
+def envelope(z: np.ndarray, bp_in: tuple[int | None, int | None] = (1, None), *,
+             n_out: int | None = None, squared: bool = False,
+             residual: Literal['lowpass', 'all', None] = 'lowpass',
+             axis: int = -1) -> tuple[np.ndarray, np.ndarray] | np.ndarray:
```
It would be quite nice if mutating the return type based on the inputs could be avoided, either by removing the option to have `residual=None` or by returning results object instead

DietBru replied:
I changed the return type to be a single ``np.ndarray``.

j-bowhay commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,277 @@ def hilbert2(x, N=None):
+    Parameters
+    ----------
+    z :
+        Real- or complex-valued input signal, which is assumed to be made up of `n`
+        samples and having sampling interval `T`. `z` may also be a multidimensional
+        array with the time axis being defined by `axis`.
+    bp_in :
+        2-tuple defining the frequency band ``bp_in[0]:bp_in[1]`` of the input filter.
+        The corner frequencies are specified as integer multiples of `1/(n*T)` with
+        ``-n//2 <= bp_in[0] < bp_in[1] <= (n+1)//2`` being the allowed frequency range.
+        ``None`` entries are replaced with `-n//2` or `(n+1)//2` respectively. The
+        default of ``(1, None)`` removes the mean value as well as the negative
+        frequency components.
```
single backticks should only be used to reference arguments to the function
![image](https://github.com/user-attachments/assets/79681642-72c7-42f9-af9d-cb6c4184099e)

```suggestion
        Real- or complex-valued input signal, which is assumed to be made up of ``n``
        samples and having sampling interval ``T``. `z` may also be a multidimensional
        array with the time axis being defined by `axis`.
    bp_in :
        2-tuple defining the frequency band ``bp_in[0]:bp_in[1]`` of the input filter.
        The corner frequencies are specified as integer multiples of ``1/(n*T)`` with
        ``-n//2 <= bp_in[0] < bp_in[1] <= (n+1)//2`` being the allowed frequency range.
        ``None`` entries are replaced with ``-n//2`` or ``(n+1)//2`` respectively. The
        default of ``(1, None)`` removes the mean value as well as the negative
        frequency components.
```

DietBru replied:
Fixed. I also moved the description into the Notes section to be able to use LaTeX formulas.

j-bowhay commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,277 @@ def hilbert2(x, N=None):
+    squared :
+        If set, the square of the envelope is returned. The bandwidth of the squared
+        envelope is often smaller than the non-squared envelope bandwidth due to the
+        nonlinear nature of the utilized `abs` function. I.e., the embbeded square root
```
```suggestion
        nonlinear nature of the utilized ``abs`` function. I.e., the embedded square root
```

DietBru replied:
Fixed.

j-bowhay commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,277 @@ def hilbert2(x, N=None):
+        function typically produces addiational harmonics.
+    residual :
+        This option determines what kind of residual, i.e., the signal part which the
+        input bandpass filter removes, is returned. `'all'` returns everything except
+        the contents of the frequency band ``bp_in[0]:bp_in[1]``, `'lowpass'`
+        returns the contents of the frequency band ``< bp_in[0]``. If ``None`` then
+        nothing is returned.
```
```suggestion
        input bandpass filter removes, is returned. ``'all'`` returns everything except
        the contents of the frequency band ``bp_in[0]:bp_in[1]``, ``'lowpass'``
        returns the contents of the frequency band ``< bp_in[0]``. If ``None`` then
        nothing is returned.
```

DietBru replied:
Fixed.

j-bowhay commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,277 @@ def hilbert2(x, N=None):
+    filter is used. The low-pass residuum of the input bandpass filter is utilized to
+    determine an asymmetric upper and lower bound to enclose the signal. Due to the
+    smoothness of the resulting envelope, it is down-sampled from 500 to 40 samples.
+    Note that the instantaneous amplitude `a_x` and the computed envelope `x_env` are
+    not perfectly identical. This is due to the signal not being perfectly periodic as
+    well as the existence of some spectral overlapping of `x_carrier` and `x_drift`.
+    Hence, they cannot be completely separated by a bandpass filter.
```
```suggestion
    Note that the instantaneous amplitude ``a_x`` and the computed envelope ``x_env`` are
    not perfectly identical. This is due to the signal not being perfectly periodic as
    well as the existence of some spectral overlapping of ``x_carrier` and ``x_drift``.
    Hence, they cannot be completely separated by a bandpass filter.
```

DietBru replied:
Fixed.

j-bowhay commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,277 @@ def hilbert2(x, N=None):
+    The right plot shows the real part of that analytic signal being interpreted
+    as a complex-vauled signal, i.e., having zero imaginary part. There the resulting
+    envelope is not as smooth as in the analytic case and the instantaneous amplitude
+    in the real plane is not recovered. If `z_re` had been passed as a real-valued
```
```suggestion
    in the real plane is not recovered. If ``z_re`` had been passed as a real-valued
```

DietBru replied:
Fixed.

j-bowhay commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -2800,6 +2800,188 @@ def test_hilbert2_types(self, dtype):
+    @classmethod
+    def assert_close(cls, actual, desired, msg):
+        """Little helper to compare to arrays with proper tolerances"""
+        assert_allclose(actual, desired, atol=1e-12, rtol=1e-12, err_msg=msg)
```
we should probably use `xp_assert_close` to make life easier in the future

DietBru replied:
Fixed.

j-bowhay reviewed (commented):
Thanks @DietBru a few suggestion but otherwise looks in good shape

DietBru committed:
Fix double quotes in docstring

DietBru committed:
Refactored return type and docstring.

DietBru committed:
Small docstring improvements and usage of `xp_assert_close` instead of `assert_allclose`.

DietBru committed:
More fixed double backticks.

DietBru commented:
Thanks for the feedback @j-bowhay! I think I have incorporated all your suggestions—the rendered documentation can be found [here].

[here]: https://output.circle-artifacts.com/output/job/880a3fdd-87ab-4e30-8acd-d5aee891c2a7/artifacts/0/html/reference/generated/scipy.signal.envelope.html

j-bowhay commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,282 @@ def hilbert2(x, N=None):
+
+    Returns
+    -------
+    np.ndarray
```
for consistency with above
```suggestion
    ndarray
```

DietBru committed:
Change `nd.ndarray` to `ndarray` in docstr for consistency.

Co-authored-by: Jake Bowhay <60778417+j-bowhay@users.noreply.github.com>

lucascolley commented on scipy/signal/_signaltools.py:
```diff
@@ -2327,6 +2327,10 @@ def hilbert(x, N=None, axis=-1):
 
+    See Also
+    --------
+    envelope: Compute envelope of a signal of a real- or complex-valued signal.
```
```suggestion
    envelope: Compute envelope of a real- or complex-valued signal.
```

lucascolley commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,282 @@ def hilbert2(x, N=None):
+    n_out : int | None
+        If not ``None`` the output will be resampled to `n_out` samples. The default
+        of ``None`` sets the output to the same length as the input `z`.
+    squared : bool
+        If set, the square of the envelope is returned. The bandwidth of the squared
+        envelope is often smaller than the non-squared envelope bandwidth due to the
+        nonlinear nature of the utilized absolute value function. I.e., the embedded
+        square root function typically produces addiational harmonics.
```
```suggestion
    squared : bool
        If set, the square of the envelope is returned. The bandwidth of the squared
        envelope is often smaller than the non-squared envelope bandwidth due to the
        nonlinear nature of the utilized absolute value function. I.e., the embedded
        square root function typically produces addiational harmonics.
        Default: ``False``.
```

lucascolley commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,282 @@ def hilbert2(x, N=None):
+        Real- or complex-valued input signal, which is assumed to be made up of ``n``
+        samples and having sampling interval ``T``. `z` may also be a multidimensional
+        array with the time axis being defined by `axis`.
+    bp_in : tuple[int | None, int | None]
```
```suggestion
    bp_in : tuple[int | None, int | None], optional
```

lucascolley commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,282 @@ def hilbert2(x, N=None):
+        ``None`` entries are replaced with `-n//2` or `(n+1)//2` respectively. The
+        default of ``(1, None)`` removes the mean value as well as the negative
+        frequency components.
+    n_out : int | None
```
```suggestion
    n_out : int | None, optional
```

lucascolley commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,282 @@ def hilbert2(x, N=None):
+    n_out : int | None
+        If not ``None`` the output will be resampled to `n_out` samples. The default
+        of ``None`` sets the output to the same length as the input `z`.
+    squared : bool
```
```suggestion
    squared : bool, optional
```

lucascolley commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,282 @@ def hilbert2(x, N=None):
+        envelope is often smaller than the non-squared envelope bandwidth due to the
+        nonlinear nature of the utilized absolute value function. I.e., the embedded
+        square root function typically produces addiational harmonics.
+    residual : Literal['lowpass', 'all', None]
```
```suggestion
    residual : Literal['lowpass', 'all', None], optional
```

lucascolley commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,282 @@ def hilbert2(x, N=None):
+        the contents of the frequency band ``bp_in[0]:bp_in[1]``, ``'lowpass'``
+        returns the contents of the frequency band ``< bp_in[0]``. If ``None`` then
+        only the envelope is returned.
+    axis : int
```
```suggestion
    axis : int, optional
```

lucascolley commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,282 @@ def hilbert2(x, N=None):
+        input bandpass filter removes, is returned. ``'all'`` returns everything except
+        the contents of the frequency band ``bp_in[0]:bp_in[1]``, ``'lowpass'``
+        returns the contents of the frequency band ``< bp_in[0]``. If ``None`` then
+        only the envelope is returned.
```
```suggestion
        only the envelope is returned. Default: ``'lowpass'``.
```

lucascolley commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,282 @@ def hilbert2(x, N=None):
+    Returns
+    -------
+    ndarray
+        If parameter `residual` is ``None`` then an array `z_env` with the same shape
```
```suggestion
        If parameter `residual` is ``None`` then an array ``z_env`` with the same shape
```

lucascolley commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,282 @@ def hilbert2(x, N=None):
+    ndarray
+        If parameter `residual` is ``None`` then an array `z_env` with the same shape
+        as the input `z` is returned, containing its envelope. Otherwise, an array with
+        shape ``(2, *z.shape)``, containing the arrays `z_env`, `z_res` being stacked
+        along the first axis, is returned.
```
```suggestion
        shape ``(2, *z.shape)``, containing the arrays ``z_env`` and ``z_res``, stacked
        along the first axis, is returned.
```

lucascolley commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,282 @@ def hilbert2(x, N=None):
+        shape ``(2, *z.shape)``, containing the arrays `z_env`, `z_res` being stacked
+        along the first axis, is returned.
+        It allows unpacking, i.e., ``z_env, z_res = envelope(z)``.
+        The residual `z_res` contains the signal part which the input bandpass filter
```
```suggestion
        The residual ``z_res`` contains the signal part which the input bandpass filter
```

lucascolley commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,282 @@ def hilbert2(x, N=None):
+        It allows unpacking, i.e., ``z_env, z_res = envelope(z)``.
+        The residual `z_res` contains the signal part which the input bandpass filter
+        removed, depending on the parameter `residual`. Note that for real-valued
+        signals, also a real-valued residual is returned. Hence, the negative frequency
```
```suggestion
        signals, a real-valued residual is returned. Hence, the negative frequency
```

lucascolley commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,282 @@ def hilbert2(x, N=None):
+    absolute value. Hence, it is more efficient to use ``np.abs(z)`` instead of this
+    function.
+
+    Though, computing the envelope based on the analytic signal [1]_ is the natural
+    method for real-valued signals, other methods are also frequently used. The most
```
```suggestion
    Although computing the envelope based on the analytic signal [1]_ is the natural
    method for real-valued signals, other methods are also frequently used. The most
```

lucascolley commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,282 @@ def hilbert2(x, N=None):
+
+    Though, computing the envelope based on the analytic signal [1]_ is the natural
+    method for real-valued signals, other methods are also frequently used. The most
+    popular alterantive is probably the so-called "square-law" envelope detector and
```
```suggestion
    popular alternative is probably the so-called "square-law" envelope detector and
```

lucascolley commented on scipy/signal/_signaltools.py:
```diff
@@ -2452,6 +2456,282 @@ def hilbert2(x, N=None):
+    Though, computing the envelope based on the analytic signal [1]_ is the natural
+    method for real-valued signals, other methods are also frequently used. The most
+    popular alterantive is probably the so-called "square-law" envelope detector and
+    its relatives [2]_. They do not always compute the correct result for all kind of
```
```suggestion
    its relatives [2]_. They do not always compute the correct result for all kinds of
```

lucascolley commented on scipy/signal/tests/test_signaltools.py:
```diff
@@ -2800,6 +2801,188 @@ def test_hilbert2_types(self, dtype):
+        """Little helper to compare to arrays with proper tolerances"""
+        xp_assert_close(actual, desired, atol=1e-12, rtol=1e-12, err_msg=msg)
+
+# noinspection PyTypeChecker
```
is this necessary?

lucascolley reviewed (approved):
thanks @DietBru!! Given @yagizolmez's approval, I'm happy to merge this, modulo the suggestions I have added here, and checking the rendered docs one more time. Let's get this in for 1.15.0 🎉

Please check all of my suggestions and commit them if you agree.

DietBru committed:
Update scipy/signal/_signaltools.py

Fixed typo

Co-authored-by: Lucas Colley <lucas.colley8@gmail.com>

DietBru committed:
* Fixed typos in `envelope` docstr and in class `TestEnvelope`
* In class `TestEnvelope.test_envelope_invalid_parameters()` make linter statement
  `# noinspection PyTypeChecker` expression-specific instead of method-specific

DietBru commented:
Thanks @lucascolley for the review and spotting the typos—I think, I addressed all your comments in the last commit. 
The CI errors seem to be caused by #21988 and thus unrelated. From my side it is ready for merging.

lucascolley commented:
thanks Dietrich, let's check the rendered docs once gh-21988 is resolved.

lucascolley committed:
Merge branch 'main' into EnvelopeHIlbert

[skip cirrus]

DietBru commented:
> thanks Dietrich, let's check the rendered docs once [gh-21988](https://github.com/scipy/scipy/issues/21988) is resolved.

Glancing at the [rendered docs], I saw no surprises. :tada: 

[rendered docs]: https://output.circle-artifacts.com/output/job/b5b8ec7e-4e31-44d4-994b-3e6979f531c5/artifacts/0/html/reference/generated/scipy.signal.envelope.html#scipy.signal.envelope

lucascolley committed:
Update scipy/signal/_signaltools.py

[skip ci]
